lib/using_restricted_array.rb

```diff
@@ -1,61 +1,114 @@
 # the range of 1-20 will be chosen for you.
-# All values are integers in the range of 1-221.
+# All values are integers in the range of 1-221. Chosen by test
 # RestrictedArray cannot be resized.
 
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), where n is the number of items in the array or length of the input (compare length/value of input to how many times loop runs or length)
+# Space complexity: O(1), program only stores value of i
 def length(array)
-  raise NotImplementedError
+  i = 0
+  until array[i] == nil
+    i += 1
+  end
+  return i
 end
 
 # Prints each integer values in the array
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), where n is the number of items in the array or length of the input
+# Space complexity: O(1), program only stores value of i and length
 def print_array(array)
-  raise NotImplementedError
+  length = length(array)
+  i = 0
+  while i < length
+    puts array[i]
+    i += 1
+  end
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the number of values in the array
+# Space complexity: O(1), program only stores value of i
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  i = 0
+  while i < length
+    return true if array[i] == value_to_find
+    i += 1
+  end
+  return false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the number of values in the array
+# Space complexity: O(1), program only stores value of i
 def find_largest(array, length)
-  raise NotImplementedError
+  i = 1
+  biggest_number = array[0]
+  while i < length
+    biggest_number = array[i] if array[i] > biggest_number
+    i += 1
+  end
+  return biggest_number
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the number of values in the array
+# Space complexity: O(1), program only stores value of i
 def find_smallest(array, length)
-  raise NotImplementedError
+  i = 1
+  smallest_number = array[0]
+  while i < length
+    smallest_number = array[i] if array[i] < smallest_number
+    i += 1
+  end
+  return smallest_number
 end
 
 # Reverses the values in the integer array in place
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), where n is the number of values in the array
+# Space complexity: O(1), program only stores value of i and j
 def reverse(array, length)
-  raise NotImplementedError
+  i = 0
+  j = length - 1
+
+  while i < j
+    b = array[i]
```

Comment:
Consider using descriptive names for your variables here as you did in your other methods.